cloudinit/url_helper.py

```diff
@@ -827,7 +835,14 @@ def read_url_parallel(start_time, timeout, exc_cb, log_cb):
         url = do_read_url(start_time, timeout, exception_cb, status_cb)
         if url:
             address, response = url
-            return (address, response.contents)
+            if isinstance(response, UrlError) and 503 == response.code:
+                # server isn't available, so retry
+                LOG.warning(
+                    "IMDS returned 503 error code. Retrying in %s",
+                    current_sleep_time,
+                )
+            elif isinstance(response, UrlResponse):
```

Comment:
So it looks like the old code was making some bad assumptions...particularly that `response` is a `UrlResponse` even though it could also be `UrlError` if an error occurred. It also looks like the callers of this code aren't really doing anything to handle exceptions. Previously if we did get a UrlError, the old `response.contents` would throw an error due to `contents` not being a member of UrlError. It was raising the wrong error obviously, but it still wound up having the same effect as the requests' `raise_for_status()`.

With this though, if we get an error that isn't a 503, we'll retry until we timeout, even if it's a non-recoverable error. I think we should instead raise the UrlError if it's found and not a 503.

Reply:
Good point, thanks for catching that @TheRealFalcon!

Reply:
> It looks like the old code was making some bad assumptions...particularly that response is a UrlResponse even though it could also be UrlError if an error occurred. 
> ...
> Previously if we did get a UrlError, the old response.contents would throw an error due to contents not being a member of UrlError. It was raising the wrong error obviously, but it still wound up having the same effect as the requests' raise_for_status().

This isn't correct. Previously, only UrlResponse was returned. See the change on line 755.

> It also looks like the callers of this code aren't really doing anything to handle exceptions.

Exceptions are optionally handled by the function passed in via `exception_cb`. Note the line in the original bug report:
```
2024-07-24 08:13:06,801 - DataSourceEc2.py[WARNING]: Fatal error while requesting Ec2 IMDSv2 API tokens
```

Looking into this a bit further, I think that we actually need to make a change to the function that caused this log to allow 503 exceptions to retry.